File: sdks/python/examples/basic_usage.py
"""
Basic usage example for NANDA Payments Python SDK
"""

import asyncio
from nanda_payments import NandaPaymentsClient, PaymentOptions


async def basic_example():
    """Basic payment operations example"""
    print("🐍 NANDA Payments Python SDK Example")
    print("=====================================\n")

    # Create a client using async context manager
    async with NandaPaymentsClient(
        agent_name="python-example",
        facilitator_url="http://localhost:3001",
        timeout=30.0,
    ) as client:

        # Test connection
        print("Testing connection...")
        is_connected = await client.test_connection()
        print(f"Connection: {'✅ Success' if is_connected else '❌ Failed'}\n")

        if not is_connected:
            print("⚠️  Make sure the facilitator is running on localhost:3001")
            return

        # Get supported schemes
        print("Getting supported payment schemes...")
        try:
            supported = await client.get_supported_schemes()
            print(f"Supported schemes: {len(supported.kinds)}\n")
        except Exception as e:
            print(f"Failed to get supported schemes: {e}\n")

        # Get balance (this would normally require the agent to be registered)
        print("Getting agent balance...")
        try:
            balance = await client.get_balance()
            print(f"Current balance: {balance.balance} {balance.currency}\n")
        except Exception as e:
            print(f"Could not get balance: {e}")
            print("(This is expected if the agent is not registered)\n")

        # Create a payment (without sending it)
        print("Creating a payment payload...")
        payment = await client.create_payment(
            PaymentOptions(
                amount=10,
                recipient="content-provider",
                description="Example payment"
            )
        )
        print(f"Payment created: {payment.tx_id}")
        print(f"Amount: {payment.amount} NP")
        print(f"From: {payment.from_agent}")
        print(f"To: {payment.pay_to}\n")

        # Example of making a request that would require payment
        print("Example payment request (will fail without real service)...")
        try:
            response = await client.make_payment_request(
                "https://api.example.com/premium-content",
                PaymentOptions(
                    amount=5,
                    recipient="api-provider",
                    description="API access"
                ),
                method="GET"
            )
            print(f"Response status: {response.status}")
            if response.payment_response:
                print(f"Payment completed: {response.payment_response}")
        except Exception as e:
            print(f"Payment request failed (expected): {e}")

    print("\n🎉 Example completed!")


async def agent_management_example():
    """Agent management operations example"""
    print("\n📊 Agent Management Example")
    print("============================\n")

    async with NandaPaymentsClient(
        agent_name="python-agent",
        facilitator_url="http://localhost:3001",
    ) as client:

        try:
            # Try to get agent summary
            print("Getting agent summary...")
            summary = await client.get_summary()
            print(f"Balance: {summary['balance']} NP")
            print(f"Total sent: {summary['total_sent']} NP")
            print(f"Total received: {summary['total_received']} NP")
            print(f"Transaction count: {summary['transaction_count']}")
            print(f"Last activity: {summary['last_activity'] or 'None'}\n")

        except Exception as e:
            print(f"Could not get agent summary: {e}")
            print("(This is expected if the agent is not registered)\n")

        try:
            # Try to get transaction history
            print("Getting transaction history...")
            from nanda_payments.types import TransactionQuery

            transactions = await client.get_transaction_history(
                TransactionQuery(limit=5)
            )
            print(f"Found {len(transactions)} recent transactions")
            for tx in transactions[:3]:  # Show first 3
                print(f"  {tx.tx_id}: {tx.amount_minor} NP from {tx.from_agent} to {tx.to_agent}")

        except Exception as e:
            print(f"Could not get transaction history: {e}")
            print("(This is expected if the agent has no transactions)")


async def fastapi_example():
    """Example of FastAPI integration"""
    print("\n🚀 FastAPI Integration Example")
    print("===============================\n")

    try:
        from fastapi import FastAPI, Depends
        from nanda_payments.integrations.fastapi import NandaPaymentsDependency, payment_required

        print("FastAPI integration is available!")
        print("Example usage:")
        print("""
from fastapi import FastAPI, Depends
from nanda_payments.integrations.fastapi import NandaPaymentsDependency, payment_required

app = FastAPI()

# Initialize dependency
nanda_dependency = NandaPaymentsDependency(
    agent_name="my-api",
    facilitator_url="http://localhost:3001"
)

@app.get("/premium")
async def premium_content(
    payment_verified = Depends(payment_required(10, "api-provider")),
    client = Depends(nanda_dependency)
):
    return {"message": "This is premium content!"}
        """)

    except ImportError:
        print("FastAPI is not installed. Install with: pip install fastapi")


if __name__ == "__main__":
    async def main():
        await basic_example()
        await agent_management_example()
        await fastapi_example()

    asyncio.run(main())

File: sdks/typescript/examples/basic-usage.ts
/**
 * Basic usage example for NANDA Payments SDK
 */

import { NandaPaymentsClient, createMockFacilitator } from '../src/index.js';

async function basicExample() {
  // Start a mock facilitator for testing
  const mockFacilitator = createMockFacilitator({ port: 3001 });
  await mockFacilitator.start();

  try {
    // Create a client
    const client = new NandaPaymentsClient({
      agentName: 'example-app',
      facilitatorUrl: 'http://localhost:3001',
    });

    console.log('🚀 NANDA Payments SDK Example');
    console.log('===============================\n');

    // Test connection
    console.log('Testing connection...');
    const isConnected = await client.testConnection();
    console.log(`Connection: ${isConnected ? '✅ Success' : '❌ Failed'}\n`);

    // Get supported schemes
    console.log('Getting supported payment schemes...');
    const supported = await client.getSupportedSchemes();
    console.log(`Supported schemes: ${supported.kinds.length}\n`);

    // Set initial balance for testing
    mockFacilitator.setAgentBalance('example-app', 100);

    // Get balance
    console.log('Getting agent balance...');
    const balance = await client.getBalance();
    console.log(`Current balance: ${balance.balance} ${balance.currency}\n`);

    // Simulate making a request that requires payment
    console.log('Making a payment request...');
    try {
      const response = await client.makePaymentRequest(
        'https://api.example.com/premium-content',
        {
          amount: 10,
          recipient: 'content-provider',
          description: 'Access to premium content',
        },
        {
          method: 'GET',
        }
      );

      console.log('Payment request completed!');
      console.log(`Status: ${response.status}`);
      if (response.paymentResponse) {
        console.log(`Transaction ID: ${response.paymentResponse.txId}`);
        console.log(`Amount: ${response.paymentResponse.amount} NP`);
      }
    } catch (error: any) {
      console.log(`Payment request failed: ${error.message}`);
    }

    console.log('\n🎉 Example completed successfully!');

  } finally {
    // Clean up
    await mockFacilitator.stop();
  }
}

async function agentManagementExample() {
  const mockFacilitator = createMockFacilitator({ port: 3002 });
  await mockFacilitator.start();

  try {
    const client = new NandaPaymentsClient({
      agentName: 'agent-example',
      facilitatorUrl: 'http://localhost:3002',
    });

    console.log('\n📊 Agent Management Example');
    console.log('============================\n');

    // Set up some test data
    mockFacilitator.setAgentBalance('agent-example', 500);

    // Get agent summary
    console.log('Getting agent summary...');
    const summary = await client.getSummary();
    console.log(`Balance: ${summary.balance} NP`);
    console.log(`Total sent: ${summary.totalSent} NP`);
    console.log(`Total received: ${summary.totalReceived} NP`);
    console.log(`Transaction count: ${summary.transactionCount}`);
    console.log(`Last activity: ${summary.lastActivity || 'None'}\n`);

    // Get agent info
    console.log('Getting agent information...');
    const agentInfo = await client.getAgentInfo();
    console.log(`Agent name: ${agentInfo.agent_name}`);
    console.log(`Wallet ID: ${agentInfo.walletId}`);
    console.log(`Service charge: ${agentInfo.serviceCharge}%\n`);

    // Get wallet info
    console.log('Getting wallet information...');
    const walletInfo = await client.getWalletInfo();
    console.log(`Wallet ID: ${walletInfo.walletId}`);
    console.log(`Balance: ${walletInfo.balanceMinor} ${walletInfo.currency}`);
    console.log(`Scale: ${walletInfo.scale}`);

  } finally {
    await mockFacilitator.stop();
  }
}

// Run examples
async function runExamples() {
  try {
    await basicExample();
    await agentManagementExample();
  } catch (error) {
    console.error('Example failed:', error);
    process.exit(1);
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  runExamples();
}